### PR TITLE
reset line distance when it is near max representable distance

### LINKED ISSUE #2099: dashed line bug
screen shot 2016-02-09 at 3 46 34 pm

I'm guessing this is what happens when `distance` is > than the maximum distance we can store in the buffers.

### PR DESCRIPTION
**Pull request by ansis**
fix #2099 

When the distance of the current vertex is > half the maximum distance that can be represented, reset the distance. This fixes the original problem where a line with many segments is longer than the maximum distance.

We still need to make sure we can represent the distance of the longest possible segment so we need to make two more changes:
- improve how `linesofar` is packed into the buffer so that we have two more bits
- scale the distance down before adding it to the buffer to increase the maximum distance that can be represented

I don't like how `var distance` changed to `this.distance` but `addCurrentVertex` needs to be able to reset the distance and I'm not seeing a better way to do this.

:eyes: @lucaswoj

### CODE REVIEW

File: js/data/line_bucket.js [hunk 1]
```diff
@@ -63,7 +68,7 @@ LineBucket.prototype.shaders = {
                 // The z component's first bit, as well as the sign bit is reserved for the direction,
                 // so we need to shift the linesofar.
                 '((dir < 0) ? -1 : 1) * ((dir ? 1 : 0) | ((linesofar << 1) & 0x7F))',
-                '(linesofar >> 6) & 0x7F'
+                '((linesofar >> 6) & 0xFF) - 128'
```
Comment (lucaswoj) [inline js/data/line_bucket.js]: Is this complex bit packing really necessary? If so, can you add a comment to the source about why it is necessary and how it works.

File: js/data/line_bucket.js [hunk 2]
```diff
@@ -27,6 +27,11 @@ var EXTRUDE_SCALE = 63;
 var COS_HALF_SHARP_CORNER = Math.cos(75 / 2 * (Math.PI / 180));
 var SHARP_CORNER_OFFSET = 15;
 
+// The maximum distance along a line that can be stored in the buffers.
+var MAX_LINE_DISTANCE = Math.pow(2, 14);
+var LINE_DISTANCE_SCALE = 1 / 2;
```
Comment (lucaswoj) [inline js/data/line_bucket.js]: Please add a comment clarifying the unit of this value (even if it is "extent units")

File: shaders/linepattern.vertex.glsl [hunk 3]
```diff
@@ -8,6 +8,10 @@ precision mediump float;
 // #define scale 63.0
 #define scale 0.015873016
 
+// We scale the distance before adding it to the buffers so that we can store
+// long distances for long segments. Use this value to unscale the distance.
```
Comment (lucaswoj) [inline shaders/linepattern.vertex.glsl]: This comment would be useful in `LineBucket` too

File: js/data/line_bucket.js [hunk 4]
```diff
@@ -175,8 +181,8 @@ LineBucket.prototype.addLine = function(vertices, join, cap, miterLimit, roundLi
             var prevSegmentLength = currentVertex.dist(prevVertex);
             if (prevSegmentLength > 2 * sharpCornerOffset) {
                 var newPrevVertex = currentVertex.sub(currentVertex.sub(prevVertex)._mult(sharpCornerOffset / prevSegmentLength)._round());
-                distance += newPrevVertex.dist(prevVertex);
-                this.addCurrentVertex(newPrevVertex, flip, distance, prevNormal.mult(1), 0, 0, false);
+                this.distance += newPrevVertex.dist(prevVertex) * LINE_DISTANCE_SCALE;
```
Comment (lucaswoj) [inline js/data/line_bucket.js]: Do we have to scale the distance here? Or can that be done [here](https://github.com/mapbox/mapbox-gl-js/pull/2114/files#diff-7ffec697472833439424ba5814cb0c00R71)? My intent was for all buffer packing craziness to go there.

File: js/data/line_bucket.js [hunk 5]
```diff
@@ -377,6 +383,14 @@ LineBucket.prototype.addCurrentVertex = function(currentVertex, flip, distance,
     }
     this.e1 = this.e2;
     this.e2 = this.e3;
+
+    // There is a maximum "distance along the line" that we can store in the buffers.
+    // When we get close to the distance, reset it to zero and add the vertex again with
+    // a distance of zero.
```
Comment (lucaswoj) [inline js/data/line_bucket.js]: Please clarify that this maximum distance is determined by the number of bits in the `linesofar` attribute

**Comment by lucaswoj** [comment]
> I don't like how `var distance` changed to `this.distance` but `addCurrentVertex` needs to be able to reset the distance and I'm not seeing a better way to do this.

I can live with this for now.

(Though I acutely feel the pain of this class's overcomplicated control flow and too-long methods)

**Comment by ansis** [comment]
I added comments, moved LINE_DISTANCE_SCALE and added a couple more simplifications: 3b5bc9e 75ccf64

I closed this pr and opened a new one against master in https://github.com/mapbox/mapbox-gl-js/pull/2159

**Pull request closed without merging**